theano/tensor/nnet/abstract_conv.py

```diff
@@ -2612,12 +2721,28 @@ def perform(self, node, inp, out_):
         o, = out_
 
         mode = self.border_mode
-        if not ((isinstance(mode, tuple) and min(mode) >= 0) or
-                mode in ('valid', 'full', 'half')):
+        if isinstance(mode, tuple):
+            if len(mode) != 2:
+                raise ValueError(
+                    'invalid border_mode {} which must be a '
+                    'tuple of length {}'.format(mode, self.convdim))
+            border = ()
+            for m in mode:
+                if isinstance(m, integer_types) and m >= 0:
+                    border += ((m, m),)
+                elif isinstance(m, tuple) and len(m) == 2 and \
+                        min(m) >= 0:
+                    border += ((int(m[0]), int(m[1])),)
+                else:
+                    raise ValueError(
+                        'invalid border mode {}. The tuple can only contain '
+                        'integers or tuples of length 2'.format(mode))
+            mode = border
+        elif mode not in ('valid', 'full', 'half'):
             raise ValueError(
                 'invalid border_mode {}, which must be either '
-                '"valid", "full", "half", an integer or a tuple of'
-                ' integers'.format(mode))
+                '"valid", "full", "half", an integer or a tuple '
+                'of length {}'.format(mode, self.convdim))
```

Comment:
This if can be moved a few lines down, and then it looks very similar to that of the other two `perform` functions, so it can be extracted into a shared helper function that computes `pad` given `border_mode`.